lib/datainterfaces/bsdi_betarislive.c

```diff
@@ -0,0 +1,200 @@
+static bgpstream_data_interface_option_t options[] = {
+  /* Firehose Client */
+  {
+    BGPSTREAM_DATA_INTERFACE_BETARISLIVE, // interface ID
+    OPTION_CLIENT,                   // internal ID
+    "client",                        // name
+    "client name for RIS-Live firehose stream (default: " DEFAULT_CLIENT ")",
+  },
+};
+
+/* create the class structure for this data interface */
+BSDI_CREATE_CLASS_FULL(
+  betarislive, "beta-ris-stream", BGPSTREAM_DATA_INTERFACE_BETARISLIVE,
+  "Read updates in real-time from the RIPE RIS live stream (BETA)",
+  options);
+
+/* ---------- END CLASS DEFINITION ---------- */
+
+typedef struct bsdi_betarislive_state {
+  /* user-provided options: */
+
+  // RIS live firehose client name
+  char *client_name;
+
+  // RIS live firehose full url
+  char *url;
+
+  // we only ever yield one resource
+  int done;
+
+} bsdi_betarislive_state_t;
+
+/* ========== PRIVATE METHODS BELOW HERE ========== */
+
+static int build_url(bsdi_t *di)
+{
+  if(STATE->client_name == NULL){
+    // assign default client name
+    if((STATE->client_name = strdup(DEFAULT_CLIENT)) == NULL){
+      bgpstream_log(BGPSTREAM_LOG_ERR,
+                    "Could not assign default ris-live firehose client.");
+      return -1;
+    }
+  }
+
+  size_t needed = snprintf(NULL, 0, "%s&client=%s", FIREHOSE_URL, STATE->client_name) + 1;
+  STATE->url = malloc(needed);
+  int written = sprintf(STATE->url, "%s&client=%s", FIREHOSE_URL, STATE->client_name);
+  if(needed != written+1){
+    return -1;
+  }
+  return 0;
+}
+
+
+/* ========== PUBLIC METHODS BELOW HERE ========== */
+
+int bsdi_betarislive_init(bsdi_t *di)
+{
+  bsdi_betarislive_state_t *state;
+
+  if ((state = malloc_zero(sizeof(bsdi_betarislive_state_t))) == NULL) {
+    goto err;
+  }
+  BSDI_SET_STATE(di, state);
+
+  return 0;
+
+err:
+  bsdi_betarislive_destroy(di);
+  return -1;
+}
+
+int bsdi_betarislive_start(bsdi_t *di)
+{
+  // our defaults are sufficient to run
+  return 0;
+}
+
+int bsdi_betarislive_set_option(
+  bsdi_t *di, const bgpstream_data_interface_option_t *option_type,
+  const char *option_value)
+{
+  int i;
+  int found = 0;
+
+  switch (option_type->id) {
+  case OPTION_CLIENT:
+    free(STATE->client_name);
+    if ((STATE->client_name = strdup(option_value)) == NULL) {
+      return -1;
+    }
+    break;
+
+  default:
+    return -1;
+  }
+
+  return 0;
+}
+
+void bsdi_betarislive_destroy(bsdi_t *di)
+{
+  if (di == NULL || STATE == NULL) {
+    return;
+  }
+
+  free(STATE->client_name);
+  STATE->client_name = NULL;
+
+  free(STATE->url);
+  STATE->url = NULL;
+
+  free(STATE);
+  BSDI_SET_STATE(di, NULL);
+}
+
+int bsdi_betarislive_update_resources(bsdi_t *di)
+{
+  int rc;
+  bgpstream_resource_t *res = NULL;
+
+  // we only ever yield one resource
+  if (STATE->done != 0) {
+    return 0;
+  }
+  STATE->done = 1;
+
+  // construct url
+  if(build_url(di) != 0){
+    return -1;
+  }
+
+  // we treat kafka as having data from <recent> to <forever>
```

Comment:
comments in this block seem to be copy/pasted

Reply:
fixed in 61253bc